fix(lib/runtime): avoid caching version in runtime instance

FlorianFranzen opened the pull request:
## Changes

- No longer provide workaround for runtimes with old babe implementation
- Stop caching runtime version inside life and wasmer instance

-> The runtime environment no longer expect the `Version()` Runtime API to be present in all runtimes.

## Tests

I have not tested this well yet, beyond checking if it fixes w3f/polkadot-tests#19.

## Issues

- w3f/polkadot-tests#19

## Primary Reviewer

Anyone

qdm12 commented on lib/runtime/life/instance.go:
```diff
@@ -41,9 +41,8 @@ type Config struct {
 
 // Instance represents a v0.8 runtime life instance
```
Should we amend this comment? :thinking: Since the version isn't contained in the instance and also it wasn't fixed :thinking:

FlorianFranzen replied:
Indeed, this should probably be updated.

qdm12 commented on lib/runtime/life/exports.go:
```diff
@@ -141,12 +140,6 @@ func (in *Instance) ExecuteBlock(block *types.Block) ([]byte, error) {
 
 	// remove seal digest only
 	for _, d := range block.Header.Digest.Types {
-		// hack since substrate node_runtime can't seem to handle BABE pre-runtime digests
-		// with type prefix (ie Primary, Secondary...)
-		if bytes.Equal(in.version.SpecName(), []byte("node")) {
-			break
-		}
```
Do we still need this?

If so, should we then keep it and change it to query the version from the runtime and do this if condition check?

FlorianFranzen replied:
The unit test currently seems to indicate so, probably because the `gssmr` runtime is not secondary slot aware.

noot replied:
I agree, I'm currently working on updating the host API and dev/gssmr runtimes to be compatible with polkadot v0.9.17. if we merge this, it will temporarily break the unit test but we can just skip it until the runtime is updated

qdm12 replied:
Which test should we skip? The `t.Skip("reason")` should be part of this PR.

timwu20 replied:
`TestInstance_ExecuteBlock_GossamerRuntime`

qdm12 reviewed (approved):
Nice, thank you!

I wonder why we cached the version initially as well :thinking:

jimjbrettj reviewed (approved):
lgtm

qdm12 commented on lib/runtime/life/instance.go:
```diff
@@ -41,9 +41,8 @@ type Config struct {
 
 // Instance represents a v0.8 runtime life instance
```
```suggestion
// Instance is a runtime life instance
```

FlorianFranzen commented:
Update the PR last night. Test is now disabled and everything passes again. (The CI is still angry I removed the "primary reviewer" section.)

FlorianFranzen commented:
> Can you do `git commit --amend --no-edit && git push --force` to re-trigger it please? 

TIL that the commit timestamp is part of the git hash...

FlorianFranzen commented:
> I've added the `Primary reviewer` section to the PR body.

Looks like the bot was still not happy with that, as it expects "Primary __R__eviewer". So I fixed that and ran the CI again.

qdm12 commented:
Merged! :tada: 

> TIL that the commit timestamp is part of the git hash...

TIL You can actually amend the commit and re-use the older timestamp using

```
git commit --amend --no-edit --date="$(git log -n 1 --format=%aD)"
```

But you still get a different commit hash :shrug: 

> Looks like the bot was still not happy with that

Yeah we may remove that CI workflow, it's rather cumbersome, sorry about that.
